### PR TITLE
Report matrices and arrays

### PR DESCRIPTION
**Pull request by 11rchitwood**
# Description

This PR adds the ability to report matrices and arrays by internally coercing using `as.data.frame()`.

# Proposed Changes

I added `report.matrix()` and `report.array()` and snapshot tests to the `report.data.frame()` test file. Maybe the two new source files should be merged into one? I wasn't sure of the best way to organize this.

### CODE REVIEW

File: R/report.array.R [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+                          range = TRUE,
+                          distribution = FALSE,
+                          levels_percentage = "auto",
+                          digits = 2,
+                          n_entries = 3,
+                          missing_percentage = "auto",
+                          ...) {
+  x <- as.data.frame(x)
+  report(x)
+}
```
Comment (IndrajeetPatil) [inline R/report.array.R:30]: Currently, none of the arguments are passed to the underlying method. I _think_ you can instead do:

```suggestion
report.array <- function(x, ...) {
  x <- as.data.frame(x)
  report(x, ...)
}
```

Can you experiment and see if that works?

**Review by IndrajeetPatil — CHANGES_REQUESTED** [review]

File: R/report.array.R (hunk 1 above)
Comment (11rchitwood) [reply R/report.array.R]: Done.

**Comment by DominiqueMakowski** [comment]
I'm not entirely convinced about that: matrices are originally quite different conceptually from dataframes, in that they contain info of the same type. So I'm not sure it's appropriate to report them like a data frame, column by column.

Take a correlation matrix, or a pairwise distance matrix... you could be interested in the average correlation/distance, the range and its global distribution, but it doesn't make much sense to describe them column by column.

I'd tend to say: matrices should be kept as matrices and reported globally by their type (i.e. as numeric if they are numeric), and if the matrix is actually a dataframe (since some R functions return matrices), then we could leave it to the user to convert it explicitly, since it would be the right thing to do anyway

what do you think?

**Comment by 11rchitwood** [comment]
Would adding a margin option be helpful? Internally, it could do something like this.
``` r
apply(WorldPhones, 1, report:::report.numeric)
#>                                                                                                                                                             1951 
#>   "Continuous variable: n = 7, Mean = 10642.00, SD = 17309.22, Median = 1815.00, MAD = 1868.08, range: [89, 45939], Skewness = 1.85, Kurtosis = 2.95, 0 missing" 
#>                                                                                                                                                             1956 
#>  "Continuous variable: n = 7, Mean = 14599.86, SD = 22712.46, Median = 2568.00, MAD = 2720.57, range: [733, 60423], Skewness = 1.80, Kurtosis = 2.64, 0 missing" 
#>                                                                                                                                                             1957 
#>  "Continuous variable: n = 7, Mean = 15714.43, SD = 24362.16, Median = 2695.00, MAD = 2849.56, range: [773, 64721], Skewness = 1.78, Kurtosis = 2.56, 0 missing" 
#>                                                                                                                                                             1958 
#>  "Continuous variable: n = 7, Mean = 16914.14, SD = 25790.60, Median = 2845.00, MAD = 2978.54, range: [836, 68484], Skewness = 1.75, Kurtosis = 2.40, 0 missing" 
#>                                                                                                                                                             1959 
#>  "Continuous variable: n = 7, Mean = 17828.71, SD = 27116.58, Median = 3000.00, MAD = 3097.15, range: [911, 71799], Skewness = 1.73, Kurtosis = 2.28, 0 missing" 
#>                                                                                                                                                             1960 
#> "Continuous variable: n = 7, Mean = 19101.29, SD = 28712.04, Median = 3145.00, MAD = 3168.32, range: [1008, 76036], Skewness = 1.71, Kurtosis = 2.19, 0 missing" 
#>                                                                                                                                                             1961 
#> "Continuous variable: n = 7, Mean = 20242.86, SD = 30213.56, Median = 3338.00, MAD = 3353.64, range: [1076, 79831], Skewness = 1.69, Kurtosis = 2.05, 0 missing"
apply(WorldPhones, 2, report:::report.numeric)
#>                                                                                                                                                               N.Amer 
#> "Continuous variable: n = 7, Mean = 66747.57, SD = 11277.46, Median = 68484.00, MAD = 11196.60, range: [45939, 79831], Skewness = -0.99, Kurtosis = 1.13, 0 missing" 
#>                                                                                                                                                               Europe 
#>   "Continuous variable: n = 7, Mean = 34343.43, SD = 7195.62, Median = 35218.00, MAD = 7595.36, range: [21574, 43173], Skewness = -0.77, Kurtosis = 0.60, 0 missing" 
#>                                                                                                                                                                 Asia 
#>      "Continuous variable: n = 7, Mean = 6229.29, SD = 2124.21, Median = 6662.00, MAD = 2309.89, range: [2876, 9053], Skewness = -0.28, Kurtosis = -0.53, 0 missing" 
#>                                                                                                                                                               S.Amer 
#>         "Continuous variable: n = 7, Mean = 2772.29, SD = 496.69, Median = 2845.00, MAD = 410.68, range: [1815, 3338], Skewness = -1.22, Kurtosis = 2.01, 0 missing" 
#>                                                                                                                                                              Oceania 
#>         "Continuous variable: n = 7, Mean = 2625.00, SD = 523.06, Median = 2691.00, MAD = 481.84, range: [1646, 3224], Skewness = -1.06, Kurtosis = 1.39, 0 missing" 
#>                                                                                                                                                               Africa 
#>           "Continuous variable: n = 7, Mean = 1484.00, SD = 647.71, Median = 1663.00, MAD = 358.79, range: [89, 2005], Skewness = -2.12, Kurtosis = 4.94, 0 missing" 
#>                                                                                                                                                             Mid.Amer 
#>           "Continuous variable: n = 7, Mean = 841.71, SD = 176.12, Median = 836.00, MAD = 152.71, range: [555, 1076], Skewness = -0.32, Kurtosis = -0.20, 0 missing"
apply(iris3, 3, report:::report.numeric)
#>                                                                                                                                                         Setosa 
#> "x: n = 200, n = 200, n = 200, n = 200, Mean = 2.54, SD = 1.85, Median = 2.10, MAD = 2.52, range: [0.10, 5.80], Skewness = 0.12, Kurtosis = -0.25, 0% missing" 
#>                                                                                                                                                     Versicolor 
#>       "x: n = 200, n = 200, n = 200, n = 200, Mean = 3.57, SD = 1.76, Median = 3.30, MAD = 2.37, range: [1, 7], Skewness = 0.11, Kurtosis = -0.53, 0% missing" 
#>                                                                                                                                                      Virginica 
#>  "x: n = 200, n = 200, n = 200, n = 200, Mean = 4.29, SD = 1.92, Median = 4.15, MAD = 2.59, range: [1.40, 7.90], Skewness = 0.12, Kurtosis = 0.03, 0% missing"
```

<sup>Created on 2022-08-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by 11rchitwood** [comment]
But for my use case, coercing to `data.frame` and then calling `report()` works. So should we close this PR and issue and table for another time?

In general, is supporting most base R stuff (including stats package) within the scope of this package?

**Pull request closed without merging**